### ISSUE TITLE
[BUG] Incorrect computation of the supervised loss in train_supervisor of TimeGAN.

### ISSUE THREAD

**Issue by gmbss0** [ydataai/ydata-synthetic#69]
**Incorrect computation of the supervised loss in train_supervisor of TimeGAN.**

I believe there is a mistake in the way the supervised loss  g_loss_s is computed inside train_supervisor. You compute:
`g_loss_s = self._mse(h[:, 1:, :], h_hat_supervised[:, 1:, :])` . 
This aligns with earlier commits of the original paper code. However, this step would merely learn an identity mapping. 

Based on the paper and the [new commits](https://github.com/jsyoon0823/TimeGAN) of the authors, I believe the loss should be computed as follows:
`g_loss_s = self._mse(h[:, 1:, :], h_hat_supervised[:, :-1, :])` . This should help learning the temporal dependencies between time steps and help guiding the generator. 

Finally, I have a question regarding the gradients of the generator for the supervised training. In the second step of training TGAN, the supervisor and generator are optimised on g_loss_s. However, the generator is not contributing to the computations and thus has no gradients. Do you have an explanation for this?

**Comment by donseca** [comment]
As far as I've understood, the first two steps pertain to the training of the autoencoding components (embedding, recovery) in terms of their ability to encode into latent space and learning temporal dependencies. Thus some sort of pre-training (Chapter 4.1 in the original paper)

What remains unclear to me is why the g_loss_s is included as is in the train_supervisor function, but the generator_loss_supervised is multiplied by 0.1 in the train_embedder function and by 100 in train_generator (while taking the square root). Any insights on this?

Agree on the change of the g_loss_s (and this should also be valid for generator_loss_supervised I guess).

**Comment by gmbss0** [comment]
I believe that all hyperparameters are purely empiracally identified parameters that seem to perform well on the datasets used by the authors. The relative weight (0.1 or 100*sqrt) may be used to balance the individual loss components. g_loss_s is added to two cross entropy components (G_loss_U and G_loss_Ue) for the generator loss, as opposed to the embedding loss. Different magnitudes of the loss components may require different scaling of g_loss_s. At least this is my intuition...Hope this helps!

**Comment by donseca** [comment]
Thanks! Good point on the cross entropy...I guess I'll leave it at intuition for now.
